docs/design/rosbag2_record_replay_service.md

```diff
@@ -0,0 +1,73 @@
+# Rosbag2 record and replay service
+
+## Context
+
+In [ros2/rosbag2#773](https://github.com/ros2/rosbag2/issues/773), it was recommended that record service requests feature should be added to rosbag2.  
+
+This design describe how to implement this feature based on current architecture and how to extend current command parameter.  
+
+## How to save record service message
+
+Service introspection ([ros2/ros2#1285](https://github.com/ros2/ros2/issues/1285)) had been implemented. All service messages (requests and responses) are published to service introspection topic which is hidden topic named as `_service_event` under service name (e.g. /add_two_ints/_service_event). So all messages in this topic can be recorded like general topic. 
+
+## How to distinguish service introspection topic in all recorded topics 
+
+Service introspection is saved like general topic. But they should use different process after reading from recording file. So they should be distinguished by topic name.  
+
+- General topic  
+  
+  [NameSpace/]TopicName
+
+- Hidden topic  
+  
+  [NameSpace/]_TopicName
+
+- Service introspection topic
+  
+  [NameSpace/]ServiceName/_service_event
+
+So if the string ends with `/_service_event` in topic name, it should be service introspection topic. Of course, it is possible that hidden topic has the same name. So message type is used for further check. The message type for service introspection must ends with "_Event" (e.g. example_interfaces/srv/AddTwoInts_Event). Therefore, service introspection topic can be identified.
+
+### Expand the 'record' command
+
+Add 1 parameter.
+
+- `-S [SerivceName1 ServiceName2 ...]` or `--services [SerivceName1 ServiceName2 ...]`
+  
+    If the user doesn't specify service names, all service introspection topics are recorded.
+
+### Change output of 'Info' command
+
+Topic for service introspection should be listed distinctively, as below example.
+
+```
+Files:             123_0.mcap
+Bag size:          28.2 KiB
+Storage id:        mcap
+Duration:          15.59s
+Start:             May 19 2023 13:22:25.340 (1684473745.340)
+End:               May 19 2023 13:22:40.400 (1684473760.400)
+Messages:          60
+Topic information: Topic: /chatter | Type: std_msgs/msg/String | Count: 16 | Serialization Format: cdr
+                   Topic: /events/write_split | Type: rosbag2_interfaces/msg/WriteSplitEvent | Count: 0 | Serialization Format: cdr
+                   Topic: /parameter_events | Type: rcl_interfaces/msg/ParameterEvent | Count: 0 | Serialization Format: cdr
+                   Topic: /rosout | Type: rcl_interfaces/msg/Log | Count: 44 | Serialization Format: cdr
+
+>>>> The below is new added items for service introspection <<<<<<<
+Service Events: XX  
+Service information: Service: /xxx/xxx | Type: xxx/xxx/xxx | Event Count: XX | Serialization Format: XX
+                     Service: /xxx/xxx | Type: xxx/xxx/xxx | Event Count: XX | Serialization Format: XX
```

Comment:
this incudes everything as `event`? i would expect that it can show `request` and `response` separately, so that user can understand how many requests and responses are recorded via `info` command?

Reply:
When retrieving information from requests and responses, it requires deserializing the recorded message and parsing the data inside.   
This process can slow down the "info" command when dealing with large recording files (record many service events).   
However, we can consider adding a new parameter that allows users to specify whether they want to retrieve detailed service information. What do you think?

Reply:
I think it makes sense to include the number of request/responses by default. I can see it being useful for users who might want to check if a bag includes requests or reponses, or if they want to investigate dropped responses.  Plus, `ros2 bag info` isn't the most performance critical command anyways

Reply:
> I think it makes sense to include the number of request/responses by default. 

Okay. 

> This process can slow down the "info" command when dealing with large recording files (record many service events).
However, we can consider adding a new parameter that allows users to specify whether they want to retrieve detailed service information. What do you think?

Sorry. I forgot that metadata file was created after recording. Typically, this file exists, which eliminates the need for re-analyzing the recording file. That is, info can be parsed from this file.

Reply:
Yes, we can keep metadata of `num_requests, num_responses` at recording time, and store that as metadata about the topic

Reply:
If we want to get num_requests and num_responses, we have to deserilize service event message to parse these information during recording. In order to avoid affecting performance, we plan to not get num_requests and num_responses while recording. Instead, if user execut info command with `-v`', num_requests and num_responses will be parsed from recorded file.

Reply:
the option sounds okay, but i am included to have this information in default. i am with https://github.com/ros2/rosbag2/pull/1359/files#r1211126103 for default behavior. after all, either is fine to me.

Reply:
Getting the number of request and response need to spend time to parse bag file.   
If bag file is big and users only want to check topic information, users have to wait parsing service information.   
So not set as default mode.